openprescribing/matrixstore/build/update_bnf_map.py

```diff
@@ -0,0 +1,122 @@
+"""
+Update the prescribing data in a SQLite file using the `bnf_map` table in
+BigQuery which maps old BNF codes to their current versions
+"""
+import logging
+import os.path
+import sqlite3
+
+from gcutils.bigquery import Client
+from matrixstore.matrix_ops import sparse_matrix, finalise_matrix, is_integer
+from matrixstore.serializer import deserialize, serialize_compressed
+
+
+logger = logging.getLogger(__name__)
+
+
+def update_bnf_map(sqlite_path):
+    if not os.path.exists(sqlite_path):
+        raise RuntimeError('No SQLite file at: {}'.format(sqlite_path))
+    connection = sqlite3.connect(sqlite_path)
+    # Disable the sqlite module's magical transaction handling features because
+    # we want to use our own transactions below
+    connection.isolation_level = None
+    cursor = connection.cursor()
+    bigquery_connection = Client('hscic')
+    bnf_map = get_old_to_new_bnf_codes(bigquery_connection)
+    for old_code, new_code in bnf_map:
+        move_values_from_old_code_to_new(cursor, old_code, new_code)
+    connection.commit()
+    connection.close()
+
+
+def get_old_to_new_bnf_codes(bigquery_connection):
+    result = bigquery_connection.query(
+        'SELECT former_bnf_code, current_bnf_code FROM {hscic}.bnf_map'
+    )
+    rows = result.rows
+    logger.info('Applying %s BNF code updates', len(rows))
+    return rows
+
+
+def move_values_from_old_code_to_new(cursor, old_code, new_code):
+    """
+    Move prescribing data stored under `old_code` to `new_code`
+
+    If we have data stored under both the old and new codes then sum them.
+    """
```

Comment:
I've tweaked the log messages a bit to try to make things clearer, but I think it would be easiest to talk about this in person as it's hard to know which bits of your own code are obvious and which aren't!